### PR TITLE
Add a navigation history recorder example

### PR DESCRIPTION
**Pull request by kmadsen**
## Description

To help debug issues, we typically need a history file. I looked at creating a production version of this to add to core, but the configuration is complex. Going to at least start with a code snippet that makes it easy for someone to get started. It's also the tool I use locally.

- [ ] I have added any issue links
- [ ] I have added all related labels (`bug`, `feature`, `new API(s)`, `SEMVER`, etc.)
- [ ] I have added the appropriate milestone and project boards

output

## Testing

Please describe the manual tests that you ran to verify your changes

- [x] I have tested locally (including `SNAPSHOT` upstream dependencies if needed) through testapp/demo app and run all activities to avoid regressions
- [x] I have tested via a test drive, or a simulation/mock location app
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes

## Checklist

- [ ] My code follows the style guidelines of this project
- [ ] I have performed a self-review of my own code
- [ ] I have updated the `CHANGELOG` including this PR
- [ ] We might need to update / push `api/current.txt` files after running `$> make core-update-api` (Core) / `$> make ui-update-api` (UI) if there are changes / errors we're 🆗 with (e.g. `AddedMethod` changes are marked as errors but don't break SemVer) 🚀 If there are SemVer breaking changes add the `SEMVER` label. See [Metalava](https://github.com/mapbox/mapbox-navigation-android/blob/master/docs/metalava.md) docs

### CODE REVIEW

File: examples/src/main/java/com/mapbox/navigation/examples/core/ReplayHistoryActivity.kt [hunk 1]
```diff
@@ -107,8 +109,11 @@ class ReplayHistoryActivity : AppCompatActivity() {
                 style,
                 mapboxNavigation,
                 navigationMapboxMap,
-                mapboxReplay
+                mapboxReplay,
+                HistoryRecorder(mapboxNavigation)
             )
+            navigationContext.historyRecorder.startRecording()
```
Comment (Guardiola31337) [inline examples/src/main/java/com/mapbox/navigation/examples/core/ReplayHistoryActivity.kt:115]: Are we recording history files from history files replayed? 🤯 

Jokes aside, I think it'd be great to have an option in the app's settings so the examples that integrate the `HistoryRecorder` will enable / disable it.

Also, `HistoryRecorder` stores history files in `com.mapbox.navigation.examples > cache > history-cache` it'd be great that those show up as well in the `Select history` list.

Comment (kmadsen) [reply examples/src/main/java/com/mapbox/navigation/examples/core/ReplayHistoryActivity.kt:115]: > Are we recording history files from history files replayed? 🤯
> Jokes aside, I think it'd be great to have an option in the app's settings so the examples that integrate the HistoryRecorder will enable / disable it.

haha yeah only adding it here for an example, but you're right that is not practical. Good suggestion to add it to the settings, I'll look into it. Also good suggestion to make it show up in select history 👍

Comment (kmadsen) [reply examples/src/main/java/com/mapbox/navigation/examples/core/ReplayHistoryActivity.kt:115]: Hey look it's already there! 🙃  🤔 

Screen Shot 2020-08-31 at 10 38 57 AM

**Comment by kmadsen** [comment]
> Took this for a quick spin and run into the following crash 💥

Hmm yeah it was crashing when the history is "{}". The navigator will give empty history unless a few conditions are met. I fixed it by moving the start recording to be after MapboxNavigation.startTripSession()

https://github.com/mapbox/mapbox-navigation-android/pull/3514/files#diff-48f0ec05661f0553e084795230149e28R259-R262

**Comment by Guardiola31337** [comment]
It turns out that the issue is to call `MapboxNavigation#retrieveHistory` out of the Main thread.

**Review by Guardiola31337 — COMMENTED** [review]
Thanks for addressing the feedback @kmadsen 

One last comment, supposedly I'm recording a guided session but when I replay it, it's replayed as a _Free Drive_ one 🤔

**Comment by kmadsen** [comment]
> One last comment, supposedly I'm recording a guided session but when I replay it, it's replayed as a Free Drive one 🤔

Oh yeah, we still need to support setRoute 😅 . The example uses the initial_route that is custom to only some apps

**Review by Guardiola31337 — APPROVED** [review]
:shipit:

**Pull request merged**